sql/src/main/java/io/crate/operation/projectors/sharding/BulkShardCreationLimiter.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.projectors.sharding;
+
+import io.crate.analyze.NumberOfReplicas;
+import io.crate.executor.transport.ShardRequest;
+import io.crate.metadata.PartitionName;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.cluster.metadata.IndexTemplateMetaData;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.logging.Loggers;
+
+import java.util.function.Predicate;
+
+public class BulkShardCreationLimiter<TReq extends ShardRequest<TReq, TItem>, TItem extends ShardRequest.Item>
+    implements Predicate<ShardedRequests<TReq, TItem>> {
+
+    static final int MAX_NEW_SHARDS_PER_NODE = 10;
+    private static final Logger LOGGER = Loggers.getLogger(BulkShardCreationLimiter.class);
+
+    private final ClusterService clusterService;
+
+    BulkShardCreationLimiter(ClusterService clusterService) {
```

Comment:
Would it make sense to reduce the dependency here - to just the ClusterState or even templates maybe? - depending what information is available on the use site we might even know the table/partition name up-front?

I think we could neglect the fact that the clusterState / number of replicas could change *during* a operation. That shouldn't matter for what it does.

Reply:
yes agree, wasn't sure about that, appreciate your thoughts.